File: recovery-calculator/frontend/src/lib/study.ts
/**
 * Design-study variables, and how their values resolve.
 *
 * Pure: no React, no context. Everything that needs a unit takes an injected
 * converter, the same shape `lib/corners.ts` uses, so this module is importable
 * by tests that have no provider tree and the screen still renders in whatever
 * the Units tab says.
 *
 * `physics/study.py` is the other half of the resolver. The GUI shows the
 * resolved grid and the run count BEFORE the backend ever runs, so if the two
 * disagreed the user would be shown a study that is not the one computed. They
 * are held together by a shared case table in `study.test.ts` and
 * `tests/test_study.py`.
 */

import type { UiStudyAxis, WireCanopy, WirePad } from '../types/schema'
import type { Kind } from './quantities'

/** SI value plus its quantity, to a display string. `useUnits().num`. Absent in
 *  tests, which get SI. */
export type Show = (si: number, kind: Kind) => string

/** Whether an axis lives on the vehicle, on one named device, or on the site.
 *  Mirrors `study.VEHICLE_KEYS` / `DEVICE_KEYS` / `SITE_KEYS`. */
export type Scope = 'vehicle' | 'device' | 'site'

export interface StudyVar {
  key: string
  scope: Scope
  /** What the dropdown and the table column say. */
  label: string
  /** The quantity, for variables that have one. Absent means dimensionless
   *  (Cx, n, j) or seconds (delay), which have no imperial form -- see the
   *  deliberate omissions at the top of `quantities.ts`. */
  kind?: Kind
  /** Whole numbers only. `j` is 1 or 2; `n` and the rest are continuous. */
  integer?: boolean
  /** Canopies cannot be interpolated, so no linear mode. */
  listOnly?: boolean
  help: string
}

/**
 * Everything a design study can vary, in menu order.
 *
 * Vehicle first because it is short and the device list repeats per device.
 * Within devices, the catalogue swap leads: it is the question the tab exists
 * for, and the rest are refinements on top of a chosen canopy.
 */
export const STUDY_VARS: StudyVar[] = [
  { key: 'm', scope: 'vehicle', label: 'Descending mass', kind: 'mass',
    help: 'Total mass under the canopy. Note the canopy masses come OUT of '
        + 'this, so swapping to a bigger chute does not raise it on its own.' },
  { key: 'h_a', scope: 'vehicle', label: 'Apogee', kind: 'altitude',
    help: 'How high the flight gets. Sets how long the descent has to run.' },
  { key: 'd_body', scope: 'vehicle', label: 'Airframe diameter', kind: 'length',
    help: 'Body tube diameter. Feeds the airframe drag area through eqs (14) '
        + 'and (15), so it moves the descent even before a canopy is out.' },
  { key: 'l_body', scope: 'vehicle', label: 'Airframe length', kind: 'length',
    help: 'Body tube length. Only enters through the broadside drag bound.' },

  { key: 'canopy', scope: 'device', label: 'Canopy (from catalogue)',
    listOnly: true,
    help: 'Swap the whole canopy: drag area, nominal diameter, mass and cloth '
        + 'type together, from vendor data. Everything about how it is '
        + 'deployed — trigger, delay, packing — stays as configured.' },
  { key: 'CdS', scope: 'device', label: 'Drag area (CdS)', kind: 'area',
    help: 'Full-open drag area. One atomic symbol — never a coefficient times '
        + 'an area (§1.2). Prefer the canopy swap, which keeps D0 and mass '
        + 'consistent with it.' },
  { key: 'D0', scope: 'device', label: 'Nominal diameter (D0)', kind: 'length',
    help: 'Enters only through the filling distance, eq (9) — so it changes '
        + 'how long the canopy takes to open, not how much it drags.' },
  { key: 'm_c', scope: 'device', label: 'Canopy mass', kind: 'mass',
    help: 'Canopy, lines and bag. Comes out of the descending mass, so a '
        + 'heavier canopy means a lighter body, not a heavier vehicle.' },
  { key: 'trigger', scope: 'device', label: 'Deploy trigger',
    help: 'Altitude AGL, or seconds after apogee — whichever this device is '
        + 'set to. The kind is not swept: that would be two designs, not one '
        + 'axis.' },
  { key: 'delay', scope: 'device', label: 'Deployment delay',
    help: 'Charge to line stretch, seconds. 0 free-packed, 0.3–1.0 s bagged. '
        + 'Not a rounding term: it doubles a drogue opening load at 1 s.' },
  { key: 'n', scope: 'device', label: 'Filling constant (n)',
    help: 'Canopy diameters fallen during inflation. Sets the fill time.' },
  { key: 'Cx', scope: 'device', label: 'Opening coefficient (Cx)',
    help: 'Eq (23). The dominant uncertainty in the model — worth sweeping '
        + 'here only to see its effect; the Corners tab already bounds it.' },
  { key: 'j', scope: 'device', label: 'Cloth type (j)', integer: true,
    listOnly: true,
    help: 'Area growth exponent: 2 solid cloth, 1 slotted.' },
  { key: 'v_rel', scope: 'device', label: 'Separation velocity', kind: 'speed',
    help: 'Body-to-canopy speed at line stretch. Drives the snatch load. NOT '
        + 'the descent speed.' },
  { key: 'k_eff', scope: 'device', label: 'Harness stiffness', kind: 'stiffness',
    help: 'Series stiffness of the load path, eq (32). A softer harness cuts '
        + 'the snatch load and does nothing to the opening load.' },

  // Last, because they vary the air rather than the vehicle. Both write the
  // same three site fields and the backend refuses to cross them -- see
  // `Config._check_study`.
  { key: 'pad_source', scope: 'site', label: 'Pad state (pressure source)',
    listOnly: true,
    help: 'Compare where the pad state comes from: the standard column, a pad '
        + 'barometer, the latest METAR, or the station monthly normal. Each '
        + 'point carries the temperature, pressure and lapse rate it resolved '
        + 'to, so it is a comparison of atmospheres, not of labels.' },
  { key: 'pad_month', scope: 'site', label: 'Month (monthly normals)',
    listOnly: true,
    help: 'The same station record, month by month — what the season is worth '
        + 'on descent time and impact speed. Summer air is thinner, so the '
        + 'vehicle descends faster and lands harder on the same canopy.' },
]

/** Site axes, by key. Both carry `pads` rather than numbers, and neither takes
 *  a device: there is one atmosphere over the whole flight. Mirrors
 *  `study.SITE_KEYS`. */
export const SITE_KEYS = ['pad_source', 'pad_month'] as const

export const isSiteKey = (key: string): boolean =>
  (SITE_KEYS as readonly string[]).includes(key)

export function studyVar(key: string): StudyVar | undefined {
  return STUDY_VARS.find((v) => v.key === key)
}

/**
 * The quantity an axis's values are in.
 *
 * `trigger` is the one variable whose unit is not a property of the variable:
 * a device on an ALTITUDE trigger deploys at a height and one on a TIME
 * trigger deploys at a time, and showing seconds in feet would be nonsense. So
 * it is read off the targeted device rather than the table above.
 */
export function axisKind(
  axis: { key: string; device: string | null },
  devices: { name: string; trigger: { kind: string } }[],
): Kind | undefined {
  if (axis.key === 'trigger') {
    const d = devices.find((x) => x.name === axis.device)
    return d?.trigger.kind === 'ALTITUDE' ? 'altitude' : undefined
  }
  return studyVar(axis.key)?.kind
}

/**
 * The values one axis takes, in order. SI in, SI out — units never enter here.
 *
 * Linear includes BOTH ends: `points` is how many values you get, not how many
 * gaps, so 5 points from 800 to 2000 is 800/1100/1400/1700/2000. `points === 1`
 * is the start alone, which makes pinning an axis cost a factor of one rather
 * than being a special case.
 *
 * Identical arithmetic to `study.axis_values` in Python, in the same order, so
 * the two produce bit-identical doubles rather than merely close ones.
 */
export function axisValues(axis: UiStudyAxis): (number | WireCanopy | WirePad)[] {
  if (axis.mode === 'list') {
    if (axis.key === 'canopy') return [...(axis.canopies ?? [])]
    if (isSiteKey(axis.key)) return [...(axis.pads ?? [])]
    return [...(axis.values ?? [])]
  }
  const { start, stop, points } = axis
  if (start === null || stop === null || points === null || points < 1) return []
  if (points === 1) return [start]
  const step = (stop - start) / (points - 1)
  const out: number[] = []
  for (let i = 0; i < points - 1; i += 1) out.push(start + i * step)
  // The last point is `stop` verbatim, not start + (n-1)*step. Those differ in
  // the last digit, and a top-of-range reading 1999.9999999999998 looks like a
  // bug in the physics rather than in IEEE 754.
  out.push(stop)
  return out
}

/** How many designs the study describes: the product over enabled axes. */
export function runCount(axes: UiStudyAxis[]): number {
  return axes
    .filter((a) => a.enabled)
    .reduce((n, a) => n * axisValues(a).length, 1)
}

/** Matches `study.MAX_RUNS`. The backend enforces it; this is so the button
 *  can go grey before anyone presses it. */
export const MAX_RUNS = 20

/** One axis value as text. Canopies show their label; numbers go through the
 *  injected converter when they have a quantity, and are trimmed otherwise so
 *  a table of `n` reads 6 rather than 6.000. */
export function axisValueLabel(
  value: number | string | WireCanopy | WirePad,
  kind?: Kind,
  show?: Show,
): string {
  if (typeof value === 'object') return value.label
  if (typeof value === 'string') return value
  if (kind && show) return show(value, kind)
  return String(Number(value.toFixed(4)))
}

/** What an axis is called on screen: `main · Canopy (from catalogue)`. The
 *  device name leads because with two devices it is the part that disambiguates
 *  two otherwise identical column headers. */
export function axisTitle(axis: { key: string; device: string | null }): string {
  const label = studyVar(axis.key)?.label ?? axis.key
  return axis.device ? `${axis.device} · ${label}` : label
}

File: star-openrocket/physics/study.py
"""Design studies: comparing choices, not uncertainties.

The §11.9 corner sweep in `cases.py` asks **how much of this answer is
actually known** -- it perturbs the parameters nobody has measured (Cx, n,
delay, v_rel, the airframe attitude band) to their endpoints, in common across
every device, and reports the governing corner. That is a question about
ignorance.

This asks the other question: **which design is better.** It varies things the
designer picks -- which canopy, how high the main comes out, how heavy the
vehicle ended up -- over a grid or an explicit list, applied to one named
device, and reports the four summary figures for each. Nothing here is a
bound; every point is a vehicle somebody could actually build.

The two differ in every dimension (endpoints vs a grid, in-common vs
per-device, all four §11.5 cases vs nominal only), which is why this is its own
module and its own request rather than more keys on `SweepParam`.
"""

import itertools

from pydantic import ValidationError

from physics.cases import evaluate

# The run count is the product of the axes, so it grows fast enough that a cap
# has to exist somewhere. 20 is chosen at the *display* end rather than the
# compute end: 20 integrations is under a second, but 20 lines is already more
# than one set of axes can carry, and the tab draws them all on one.
MAX_RUNS = 20

# Where each key lives. Mirrors `cases.PER_DEVICE_KEYS`/`VEHICLE_KEYS` in
# shape, so there is one place per module that says which object a key is an
# attribute of -- and `schema.StudyAxis` reads VEHICLE_KEYS from here rather
# than keeping a second copy that could disagree.
VEHICLE_KEYS = ("m", "h_a", "d_body", "l_body")
DEVICE_KEYS = ("CdS", "D0", "m_c", "j", "n", "Cx", "delay", "v_rel", "k_eff",
               "trigger", "canopy")
# The third scope. Both keys write the same three `Site` fields; what differs
# is the question -- see `StudyKey`. `Config._check_study` refuses to cross
# them for exactly that reason.
SITE_KEYS = ("pad_source", "pad_month")

# The four fields one catalogue row supplies. NOT everything on a device: n,
# Cx, delay, v_rel, k_eff and the trigger are properties of how you deploy it
# and how it was packed, not of what the vendor sells, so swapping canopies
# leaves them alone.
CANOPY_FIELDS = ("CdS", "D0", "m_c", "j")

# The three fields one pad state supplies -- §5's whole input. `z_site` is
# deliberately not among them: the project models one range, so pad elevation
# is a constant rather than a thing a study may vary.
PAD_FIELDS = ("T_pad", "p_pad", "lapse")

# Keys whose points are objects carrying their own display name, rather than
# numbers. The chart legend and the table cells read `label`; `physics/` never
# formats anything, so this is the one place that says which keys have one.
LABELLED_KEYS = ("canopy",) + SITE_KEYS


def axis_values(axis):
    """The values one axis takes, in order.

    Linear is inclusive of BOTH ends -- `points` is how many values you get,
    not how many gaps. 5 points from 800 to 2000 is 800, 1100, 1400, 1700,
    2000; a reader who asks for the range 800-2000 and gets 1760 as the top
    would have to work out why. `points == 1` is the start, which makes pinning
    an axis cost a factor of one rather than being a special case.

    `frontend/src/lib/study.ts` computes this identically, because the GUI
    shows the resolved grid before the backend ever runs. The two are held
    together by a shared case table in `tests/test_study.py` and
    `lib/study.test.ts`.
    """
    from physics.schema import StudyMode, _list_field

    if axis.mode is StudyMode.LIST:
        listed = _list_field(axis.key)
        return list(getattr(axis, listed) if listed else axis.values)

    if axis.points == 1:
        return [axis.start]
    step = (axis.stop - axis.start) / (axis.points - 1)
    # The last point is the stop VERBATIM, not start + (n-1)*step. Floating
    # point makes those differ in the last digit, and a sweep whose top value
    # reads 1999.9999999999998 in the table looks like a bug in the physics.
    return [axis.start + i * step for i in range(axis.points - 1)] + [axis.stop]


def enabled_axes(config):
    """The study's enabled axes. Empty when there is no study."""
    return [a for a in (config.study or []) if a.enabled]


def run_count(config):
    """How many designs this study describes."""
    n = 1
    for axis in enabled_axes(config):
        n *= len(axis_values(axis))
    return n


def _device(devices, name):
    for d in devices:
        if d.name == name:
            return d
    # Unreachable through the API -- Config._hard_errors checks this first --
    # but a direct library caller can construct one, and a silent no-op would
    # report a study that never varied anything.
    raise ValueError("no device named %r" % name)


def _apply(cfg, axis, value):
    """Set one axis's value on a (already-copied) config."""
    from physics.schema import StudyKey

    key = axis.key
    if key.value in VEHICLE_KEYS:
        setattr(cfg.vehicle, key.value, value)
        return

    if key.value in SITE_KEYS:
        # All three fields together, including the Nones. Copying only the
        # non-null ones would leave a monthly normal's measured lapse rate in
        # place under an ISA point, which is an atmosphere that never existed
        # -- the same failure the GUI's single month picker exists to prevent.
        for field in PAD_FIELDS:
            setattr(cfg.site, field, getattr(value, field))
        return

    device = _device(cfg.devices, axis.device)
    if key is StudyKey.canopy:
        for field in CANOPY_FIELDS:
            setattr(device, field, getattr(value, field))
    elif key is StudyKey.trigger:
        # The value only. Changing ALTITUDE to TIME would be a different
        # deployment scheme, not a point on an axis.
        device.trigger.value = value
    elif key is StudyKey.j:
        setattr(device, "j", int(value))
    else:
        setattr(device, key.value, value)


def _label(axis, value):
    """What this axis's value is called, for the point's `values` map."""
    return value.label if axis.key.value in LABELLED_KEYS else value


def _describe(values):
    """One design point in words, for an error message: `main trigger=2000`."""
    return ", ".join("%s=%s" % (k, v) for k, v in values.items())


def current_value(config, axis):
    """What this axis is set to in the config as it stands.

    So the reference row of the table can say where the current design sits on
    each axis rather than leaving the columns blank. Without it "current
    config" is a row of four numbers with nothing to attach them to, and the
    common case -- the fitted canopy IS one of the ones being compared -- is
    invisible.

    Canopies return the LABEL of whichever compared canopy the device matches,
    or None when the fitted one is not among them. The comparison is on the
    four fields the swap actually sets, because that is what "the same canopy"
    means here: two rows with the same drag area, diameter, mass and cloth type
    are the same design point whatever the vendor calls them.

    Pad states work the same way and for the same reason -- an ISA point and a
    January normal are the same design point if they resolve to the same three
    numbers, whatever the user picked them from.
    """
    from physics.schema import StudyKey

    if axis.key.value in VEHICLE_KEYS:
        return getattr(config.vehicle, axis.key.value)

    if axis.key.value in SITE_KEYS:
        for pad in axis.pads or []:
            if all(getattr(pad, f) == getattr(config.site, f)
                   for f in PAD_FIELDS):
                return pad.label
        return None

    device = _device(config.devices, axis.device)
    if axis.key is StudyKey.trigger:
        return device.trigger.value
    if axis.key is StudyKey.canopy:
        for canopy in axis.canopies or []:
            if all(getattr(canopy, f) == getattr(device, f)
                   for f in CANOPY_FIELDS):
                return canopy.label
        return None
    return getattr(device, axis.key.value)


def run_points(config, atm=None):
    """Every design in the study: [(values, CaseResult)], in product order.

    Nominal case, axial airframe -- the /api/simulate defaults. A study
    compares designs against each other, so the off-nominal cases would
    multiply the run count by four to answer a question the Recovery tab
    already answers for each design individually.

    Order is `itertools.product` over the axes in the caller's order, which is
    stable, so point 7 is the same design on every request for the same spec
    and the GUI can key colour and selection off the index.
    """
    from physics.atmosphere import Atmosphere
    from physics.schema import Config

    if atm is None:
        atm = Atmosphere(config.site.z_site, config.site.T_pad,
                         config.site.p_pad, config.site.lapse)

    axes = enabled_axes(config)
    grids = [axis_values(a) for a in axes]

    # A site axis moves the atmosphere itself, so the one built above -- from
    # the config as posted -- is only right for the points that did not move
    # it. Rebuilding per point costs a handful of flops (`Atmosphere` is an
    # eq-7 re-fit, not a table); reusing the caller's would run every month of
    # the year through January's air and report the differences as if they
    # were seasonal.
    site_swept = any(a.key.value in SITE_KEYS for a in axes)

    out = []
    for combo in itertools.product(*grids):
        cfg = config.model_copy(deep=True)
        for axis, value in zip(axes, combo):
            _apply(cfg, axis, value)

        values = {a.key.value: _label(a, v) for a, v in zip(axes, combo)}

        # Re-validate the MUTATED config, not just the one that was posted.
        # This is what makes a deploy-altitude sweep safe: `_hard_errors`
        # rejects a trigger above apogee, and without this step a study that
        # walked the main's altitude past 3000 ft would quietly produce points
        # where the device never fires -- a descent under drogue alone,
        # reported as if it were the design asked for.
        #
        # The point is named in the message. "above apogee" alone leaves the
        # reader to work out which of twenty designs is the illegal one, and
        # it is nearly always the top of exactly one axis.
        try:
            cfg = Config.model_validate(cfg.model_dump())
        except ValidationError as exc:
            raise ValueError(
                "design %s is not a valid vehicle: %s"
                % (_describe(values), "; ".join(
                    e["msg"].replace("Value error, ", "")
                    for e in exc.errors()))) from exc

        point_atm = atm
        if site_swept:
            point_atm = Atmosphere(cfg.site.z_site, cfg.site.T_pad,
                                   cfg.site.p_pad, cfg.site.lapse)

        out.append((values, evaluate(cfg, "axial", "nominal", atm=point_atm)))
    return out
